server/events/output_updater.go

```diff
@@ -0,0 +1,161 @@
+package events
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/runatlantis/atlantis/server/core/config/valid"
+	"github.com/runatlantis/atlantis/server/events/command"
+	"github.com/runatlantis/atlantis/server/events/models"
+	"github.com/runatlantis/atlantis/server/events/vcs"
+	"github.com/runatlantis/atlantis/server/events/vcs/types"
+	"github.com/runatlantis/atlantis/server/logging"
+	"github.com/runatlantis/atlantis/server/lyft/feature"
+)
+
+type OutputUpdater interface {
+	UpdateOutput(ctx *command.Context, cmd PullCommand, res command.Result)
+}
+
+// proxy that forwards req to configured output updater based on VCSHost type
+// need to add this proxy since output_updater is vcs provider agnostic
+type OutputUpdaterProxy struct {
+	outputUpdater map[models.VCSHostType]OutputUpdater
+}
+
+func (c *OutputUpdaterProxy) UpdateOutput(ctx *command.Context, cmd PullCommand, res command.Result) {
+	c.outputUpdater[ctx.HeadRepo.VCSHost.Type].UpdateOutput(ctx, cmd, res)
+}
+
+func NewOutputUpdaterProxy(pullOutputUpdater OutputUpdater, checksOutputUpdater OutputUpdater, logger logging.Logger, featureAllocator feature.Allocator, enableGithubChecks bool) OutputUpdater {
+	// All vcs host have pull output updater configured by default.
+	ouputUpdater := map[models.VCSHostType]OutputUpdater{
+		models.Github:          pullOutputUpdater,
+		models.Gitlab:          pullOutputUpdater,
+		models.AzureDevops:     pullOutputUpdater,
+		models.BitbucketCloud:  pullOutputUpdater,
+		models.BitbucketServer: pullOutputUpdater,
+	}
+
+	if enableGithubChecks {
+		ouputUpdater[models.Github] = &FeatureAwareChecksOutputUpdater{
+			checks:           checksOutputUpdater,
+			pull:             pullOutputUpdater,
+			Logger:           logger,
+			featureAllocator: featureAllocator,
+		}
+	}
+
+	return &OutputUpdaterProxy{
+		outputUpdater: ouputUpdater,
+	}
+}
+
+// defaults to pull comments if checks is turned off
+type FeatureAwareChecksOutputUpdater struct {
+	checks           OutputUpdater
+	pull             OutputUpdater
+	featureAllocator feature.Allocator
+	Logger           logging.Logger
+}
+
+func (c *FeatureAwareChecksOutputUpdater) UpdateOutput(ctx *command.Context, cmd PullCommand, res command.Result) {
+	shouldAllocate, err := c.featureAllocator.ShouldAllocate(feature.GithubChecks, ctx.HeadRepo.FullName)
+	if err != nil {
+		c.Logger.Error(fmt.Sprintf("unable to allocate for feature: %s, error: %s", feature.GithubChecks, err))
+	}
+
+	if shouldAllocate {
+		c.checks.UpdateOutput(ctx, cmd, res)
+		return
+	}
+	c.pull.UpdateOutput(ctx, cmd, res)
+}
+
+// Used to support checks type output (Github checks for example)
+type ChecksOutputUpdater struct {
```

Comment:
You can just have 1 decorator here instead of having a proxy. Something like this
```golang
type ChecksOutputUpdater struct {
        PullOutputUpdater
        TitleBuilder vcs.StatusTitleBuilder
	featureAllocator feature.Allocator
	Logger           logging.Logger
}

func (c *ChecksOutputUpdater) UpdateOutput(ctx *command.Context, cmd PullCommand, res command.Result) {
    shouldAllocate, err := c.featureAllocator.ShouldAllocate(feature.GithubChecks, ctx.HeadRepo.FullName)
	if err != nil {
		c.Logger.Error(fmt.Sprintf("unable to allocate for feature: %s, error: %s", feature.GithubChecks, err))
	}
    if ctx.HeadRepo.VCSHost.Type == models.Github && shouldAllocate {
        // github checks implementation goes here. 
    } else {
        c.PullOutputUpdater.UpdateStatus(ctx, cmd, res)
    }
}
```